[차단하기] 차단하기 비즈니스 로직 및 API 구성(#22)

Minjae-An opened the pull request:
## 작업 대상

<!-- 작업한 대상을 자세히 설명해주세요. -->

- 차단 repository(`BlockRepository`)
- 차단 query, command(`BlockQuery`, `BlockCommand`)
- 차단하기 비즈니스 로직(`BlockService` )
- 차단하기 API(`UserController` )

## 📄 작업 내용

<!-- 작업한 내용을 간단히 요약해주세요. -->
- 차단하기 비즈니스 로직 구성
    - 자기 자신을 차단하는 예외 상황 검증
    - 이미 차단한 사용자 여부 검증
    - 차단 내역 생성 및 저장
- 차단하기 API 구성

## 🙋🏻 주의 사항

<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->

## 📎 관련 이슈

## 레퍼런스

<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->

Minjae-An committed:
feat: 차단 엔티티 저장시 createdAt 자동 설정(#22)

JpaAuditing 활용하여 차단 엔티티 저장시 createdAt
필드 자동 설정되도록 코드 추가

Minjae-An committed:
feat: 차단 query repository 추가(#22)

차단 엔티티의 조회 작업을 전담하는 repository 추가

Minjae-An committed:
feat: 차단 엔티티 command repository 추가(#22)

차단 엔티티의 생성/수정/삭제 작업을 전담하는 repository 추가

Minjae-An committed:
feat: 차단하기 비즈니스 로직 구현(#22)

차단하기 비즈니스 로직은 다음 과정을 거친다.
1. 자기 자신을 차단하는 예외 상황 검증
2. 이미 차단한 사용자인지 검증
3. 차단 엔티티 생성 및 저장

Minjae-An committed:
feat: 차단하기 API 구현(#22)

Minjae-An committed:
feat: 이미 차단한 사용자 여부 조회 로직 추가(#22)

기존 차단 엔티티 조회 로직을 삭제하고, 이미 차단한 사용자
여부를 조회할 수 있는 로직을 추가

Minjae-An committed:
refactor: 이미 차단한 사용자 여부 확인 로직 수정(#22)

차단 엔티티를 조회하여 확인하는 것이 아닌 차단 엔티티의 존재
여부를 통해 확인하도록 로직 수정

Minjae-An committed:
feat: 차단 repository 정의 및 차단 존재 여부 확인 로직 추가(#22)

차단한 사용자, 차단 당한 사용자를 바탕으로 차단 존재 여부를 확인하는 로직 추가

Minjae-An committed:
refactor: 불필요한 repository 삭제(#22)

Minjae-An committed:
feat: 차단 내역 존재 확인 로직 구현(#22)

차단한 사용자, 차단 당한 사용자 기반 차단 내역이 존재하는 지
확인하는 로직 구현

Minjae-An committed:
feat: 차단 저장 로직 구현(#22)

Minjae-An committed:
feat: 차단 연관 예외 enum 정의(#22)

- 자기 자신을 차단하는 경우
- 이미 차단한 사용자인 경우

Minjae-An committed:
refactor: 차단하기 비즈니스 로직 수정(#22)

- 차단 query, command에 의존토록 수정
- 커스텀 예외 사용하

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/BlockService.java:
```diff
@@ -0,0 +1,49 @@
+      throw new GeneralException(ExceptionType.CAN_NOT_BLOCK_SELF);
+    }
+
+    User user = userService.findById(userId);
+    User blockedUser = userService.findById(blockedUserId);
```
`userService.findById()`로 할 지 `userQuery.findById()`로 할 지 논의가 필요할 것같습니다.

Minjae-An replied:
비즈니스 로직만 service에서 전담하는 구조로 가는 게 좋을 것 같습니다. `userQuery.findById()`가 적합하다고 생각됩니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/BlockService.java:
```diff
@@ -0,0 +1,49 @@
+  public void block(long userId, long blockedUserId) {
+
+    if (userId == blockedUserId) {
+      throw new GeneralException(ExceptionType.CAN_NOT_BLOCK_SELF);
```
이전 PR의 Comment와 동일한 내용으로 `IllegalArgumentException`을 사용하지 않은 이유는 무엇이고 `GeneralException`에 대해서 설명해주시면 감사하겠습니다.

acceptor-gyu reviewed (commented):
이젠 PR과 함께 논의할 내용 논의하고 설명 부탁드린 부분에 대해서 답글 달아주시면 감사하겠습니다.

Minjae-An committed:
Merge branch 'dev' into feat/#22-block

# Conflicts:
#	be/src/main/java/yeonba/be/exception/CommonException.java
#	be/src/main/java/yeonba/be/user/controller/UserController.java

Minjae-An committed:
feat: 차단 관련 예외 enum 정의(#22)

- 이미 차단한 사용자인 경우의 예외 정보 정의

Minjae-An committed:
refactor: 차단 내역 존재 확인 메서드 이름 수정(#22)

가독성 향상을 위해 간결한 이름으로 대체

Minjae-An committed:
refactor: 차단하기 비즈니스 로직 수정(#22)

- 사용자 query를 통해 사용자를 조회하도록 수정
- 자기 자신을 차단하는 경우를 사용자 엔티티가 판단하도록 수정
- 별도의 메서드로 추출했던 차단 내역 존재 확인 로직, 본 로직에 통합

Minjae-An committed:
refactor: 차단하기 API 코드 배치 수정(#22)

dev 브랜치 반영 작업 내역 병합에 따른 코드 배치 수정

Minjae-An committed:
refactor: 컨벤션에 맞게 코드 배치 수정(#22)

acceptor-gyu reviewed (approved):
굿~ 머지 시키세용~!
